Non idempotent apply terraform (#2)
nexient-llc/lcaf-component-terratest-common: README.md

```

### One-to-one relationship between examples and tests

> There would be a few scenarios where users would like to run specific tests for each example. Although this framework natively doesn't support doing that. However, as a work-around we can achieve in our `main_test.go` as follows

Reviewer: I think Philip discovered that the framework does indeed support explicitly binding test cases to example folders, check out https://github.com/nexient-llc/lcaf-component-terratest-common/blob/main/README.md?plain=1#L122-L123 and more generally his doc updates in https://github.com/nexient-llc/lcaf-component-terratest-common/commit/4fd81b54916aa6132e31ad5eb368f2fa966604a1.

Author: yes that is another way of implementing mapping between examples and tests. my approach is another. This is just a suggestion someone can choose to ignore.

Reviewer: I'm certainly not opposed to giving people options. If both are viable solutions, we should determine which solution more closely meets our needs and provide a recommendation of which to use rather than just listing both and letting the user choose without further reasoning as to why one would pick a given solution. 

There should ideally be one clear path for someone who is new to developing Terratests to take.

    // Provide a map of examples to the tests
	examplesToTestsMap := map[string]lib.TestFunc{
		"private-cluster": testimpl.TestPrivateCluster,
		"public-cluster":  testimpl.TestPublicCluster,
	}
	ctx := types.TestContext{
		TestConfig:                 &testimpl.ThisTFModuleConfig{},
		IsTerraformIdempotentApply: false,
	}
	// Loop through the examples
	for example, testFunction := range examplesToTestsMap {
		lib.RunSetupTestTeardown(t, testConfigsExamplesFolderDefault+"/"+example, infraTFVarFileNameDefault, ctx, testFunction)
	}

}
```

Currently, these tests run sequentially. Making to run them in parallel can be a future optimization.

### Run non-idempotent terraform apply
There are a few scenarios where users would like to run `terraform.initAndApply()` instead of `terraform.InitAndApplyAndIdempotent()`, mostly because of bugs in the providers which doesn't support idempotent applies. This can be done by setting the flag `IsTerraformIdempotentApply` in the context as shown below
```go
ctx := types.TestContext{
    TestConfig:                 &testimpl.ThisTFModuleConfig{},
    IsTerraformIdempotentApply: false,
}
```

### Set timeout for go test

> The default timeout of go test is 20 mins which may not be enough for running some heavy tests. Timeout can prove catastropic as it may leave resources provisioned in the cloud and cost us money. Simple way is to increase the timeout during running go tests

Reviewer: ```suggestion
The default timeout of go test is 20 mins which may not be enough for running some heavy tests. If the timeout is reached, it may leave resources provisioned in the cloud and cost us money. To increase the timeout during running go tests:
```

Author: done

```go

> go test main_test.go -timout 1h

Reviewer: ```suggestion
go test main_test.go -timeout 1h
```

Reviewer: Since we invoke `go test` from our Makefiles, do we need to propagate that change to `lcaf-component-tf-module`?

Better yet, can we detect a timeout? If it exits with a particular unique exit code, we may be able to save ourselves some headache and added cloud spend by checking `$?` and issuing a destroy command.

Author: you can pass the timeout parameter to the go test in the Makefile. may be make it configurable or a high timeout like 2 hours or so. Good question on the exit code. I dont know the answer to that question.

it will be difficult to run `terraform destroy` without the state file. I am thinking when we run the commands from the pipeline, we won't be storing the state file anywhere. it would be created and destroyed with the pipeline agent.

Reviewer: I found that the Makefile already had a timeout variable defined, so I bumped it to two hours: https://github.com/nexient-llc/lcaf-component-tf-module/pull/9

We can always set it explicitly by putting GO_TEST_TIMEOUT in our environment.

Since we're running the tests in the context of the pipeline or locally on our machines while developing, wouldn't we have the state file available to us in either scenario? If the pipeline is performing an `apply` command (via Terratest) it will have the state locally, we just need to be able to detect the condition and act upon it.

Given that you've identified this as a potential for increased cloud spend, I would like us to have either a clean solution for this that destroys resources after a timeout, or otherwise notifies us directly that this condition has occurred and someone can address it immediately, not at the end of the month when receiving the bill. I know I'm not combing through pipeline logs on the regular looking out for resources that have been deployed but not cleaned up, and I suspect nobody else is either.
